src/SignalR/common/Shared/ParseAckPipeReader.cs

```diff
@@ -0,0 +1,185 @@
+            _currentRead = res.Buffer;
+
+            if (res.IsCompleted || res.IsCanceled)
+            {
+                // TODO: figure out behavior
+                if (res.Buffer.Length >= FrameSize)
+                {
+                    res = new(res.Buffer.Slice(FrameSize), res.IsCanceled, res.IsCompleted);
+                }
+                return res;
+            }
+
+            ReadOnlySequence<byte> buffer = res.Buffer;
+            if (_remaining == 0)
+            {
+                // TODO: didn't get 24 bytes
+                var frame = buffer.Slice(0, FrameSize);
+                var len = ParseFrame(frame, _ackPipeReader);
+                _totalBytes += len;
+
+                _remaining = len;
+
+                // if the buffer doesn't have enough data we need to update how much we're slicing
+                if (len > buffer.Length - FrameSize)
+                {
+                    len = buffer.Length - FrameSize;
+                }
+
+                buffer = buffer.Slice(FrameSize, len);
+                _currentRead = buffer;
+                // 0 length means it was part of the reconnect handshake and not sent over the pipe, ignore it for acking purposes
+                // TODO: check if 0 byte writes are possible in ConnectionHandlers and possibly handle them differently
+                _ackPipeWriter.LastAck += buffer.Length == 0 ? 0 : buffer.Length + FrameSize;
+            }
+            else
+            {
+                // Advance was called and didn't consume everything even though we gave it the entire Frame Length of data
+                // This means the caller is expecting more than a single frame of data
+                // We'll need to start buffering to parse multiple frames of data
+                if (_remaining <= _currentRead.Length && buffer.Length > _remaining)
+                {
+                    // TODO: multi-frame support
+                }
+                _ackPipeWriter.LastAck += Math.Min(_remaining, newBytes);
+                _currentRead = buffer;
+                buffer = buffer.Slice(0, Math.Min(_remaining, buffer.Length));
+            }
+
+            // TODO: validation everywhere!
+            //Debug.Assert(len < res.Buffer.Length);
+
+            res = new(buffer, res.IsCanceled, res.IsCompleted);
+
+            // TODO: probably should avoid returning when we have 0 bytes to return (unless canceled/completed)
+            //Debug.Assert(buffer.Length > 0);
+        }
+        catch (Exception ex)
+        {
+            _inner.Complete(ex);
+            throw;
+        }
+
+        return res;
+    }
+
+    public static long ParseFrame(ReadOnlySequence<byte> frame, AckPipeReader ackPipeReader)
+    {
+        Debug.Assert(frame.Length >= FrameSize);
+        frame = frame.Slice(0, FrameSize);
+
+        long len;
+        long ackId;
+
+        // TODO: check perf of single Span check vs Stackalloc
```

Comment:
probably not huge for these 24-byte scenarios: https://gist.github.com/mgravell/e71e8ae48cfa607a22c4bcbe20a4b43b